contracts/test/v0.8/dev/UpkeepTranscoder3_0.test.ts

```diff
@@ -0,0 +1,449 @@
+    [
+      'tuple(uint32 paymentPremiumPPB,uint32 flatFeeMicroLink,uint32 checkGasLimit,uint24 stalenessSeconds\
+        ,uint16 gasCeilingMultiplier,uint96 minUpkeepSpend,uint32 maxPerformGas,uint32 maxCheckDataSize,\
+        uint32 maxPerformDataSize,uint256 fallbackGasPrice,uint256 fallbackLinkPrice,address transcoder,\
+        address registrar)',
+    ],
+    [config],
+  )
+}
+
+const encodeUpkeepV1 = (ids: number[], upkeeps: any[], checkDatas: any[]) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint96,address,uint32,uint64,address,uint96,address)[]',
+      'bytes[]',
+    ],
+    [ids, upkeeps, checkDatas],
+  )
+}
+
+const encodeUpkeepV2 = (ids: number[], upkeeps: any[], checkDatas: any[]) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint96,address,uint96,address,uint32,uint32,address,bool)[]',
+      'bytes[]',
+    ],
+    [ids, upkeeps, checkDatas],
+  )
+}
+
+const encodeUpkeepV3 = (
+  ids: number[],
+  upkeeps: any[],
+  checkDatas: any[],
+  admins: string[],
+) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint32,uint32,bool,address,uint96,uint96,uint32)[]',
+      'bytes[]',
+      'address[]',
+    ],
+    [ids, upkeeps, checkDatas, admins],
+  )
+}
+
+before(async () => {
+  // @ts-ignore bug in autogen file
+  upkeepTranscoderFactory = await ethers.getContractFactory(
+    'UpkeepTranscoder3_0',
+  )
+  personas = (await getUsers()).personas
+
+  linkTokenFactory = await ethers.getContractFactory('LinkToken')
+  // need full path because there are two contracts with name MockV3Aggregator
+  mockV3AggregatorFactory = (await ethers.getContractFactory(
+    'src/v0.8/tests/MockV3Aggregator.sol:MockV3Aggregator',
+  )) as unknown as MockV3AggregatorFactory
+
+  upkeepMockFactory = await ethers.getContractFactory('UpkeepMock')
+
+  owner = personas.Norbert
+  admin0 = personas.Neil
+  admin1 = personas.Nick
+  admins = [
+    (await admin0.getAddress()).toLowerCase(),
+    (await admin1.getAddress()).toLowerCase(),
+  ]
+})
+
+describe('UpkeepTranscoder3_0', () => {
+  beforeEach(async () => {
+    transcoder = await upkeepTranscoderFactory.connect(owner).deploy()
+  })
+
+  describe('#typeAndVersion', () => {
+    it('uses the correct type and version', async () => {
+      const typeAndVersion = await transcoder.typeAndVersion()
+      assert.equal(typeAndVersion, 'UpkeepTranscoder 3.0.0')
+    })
+  })
+
+  describe('#transcodeUpkeeps', () => {
+    const encodedData = '0xabcd'
+
+    it('reverts if the from type is not an enum value', async () => {
+      await evmRevert(
+        transcoder.transcodeUpkeeps(3, 2, encodedData),
+        'function was called with incorrect parameters',
+      )
+    })
+
+    it('reverts if the to version is not 2', async () => {
+      await evmRevert(
+        transcoder.transcodeUpkeeps(1, 3, encodedData),
```

Comment:
I like to re-define the enum in ts to make the tests more readable...

```ts
enum UpkeepFormat {
  V1,
  V2,
  V3
}

```

Reply:
Done